compiler/lquantizer/src/QuantizationObservers.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "QuantizationObservers.h"
+
+#include <luci/IR/CircleOpcode.h>
+
+#include <cstring>
+
+using DataType = luci_interpreter::DataType;
+
+namespace lquantizer
+{
+
+InputSavingObserver::InputSavingObserver(Node2NodeMatcher &input2node) : _input2node(input2node)
+{
+  allocate_data_buffers();
+}
+
+const std::vector<float> &InputSavingObserver::input_data_for(luci::CircleNode *node)
+{
+  assert(_node2idata.count(node));
+
+  return _node2idata[node];
+}
+
+void InputSavingObserver::allocate_data_buffers()
+{
+  // allocate input vectors for each node
+  for (auto it : _input2node)
+  {
+    auto const node = it.first;
+    auto const rank = node->rank();
+
+    // calculate number of input elements
+    auto input_elements = 1u;
+    for (int i = 0; i < rank; ++i)
+    {
+      assert(node->dim(i).known());
+      input_elements *= node->dim(i).value();
+    }
+
+    // allocate data
+    _node2idata[it.second].resize(input_elements);
+  }
```

Comment:
Iterate over `input -> node` pairs and allocate buffers in `node -> input_data` pair.